serveradmin/serverdb/query_materializer.py

```diff
@@ -136,7 +136,10 @@ def _add_attributes(self, servers_by_type):
                 for sa in ServerRelationAttribute.objects.filter(
                     value_id__in=self._server_attributes.keys(),
                     attribute_id__in=reversed_attributes.keys(),
-                ).select_related('server'):
+                ).prefetch_related('server').defer(
+                    'server__intern_ip',
+                    'server__servertype',
+                ):
```

Comment:
To make it more clear, I mean something like this (syntax and schema might be wrong, I haven't touched this in 1.5 years):

```python
from adminapi.dataset import Query

list(Query(
    {
        'servertype': 'vm',
    }, [
        'hostname',
        'intern_ip',
        {'loadbalancer': [
            'hostname',
            'intern_ip'
        ]}
    ]
))
```

Reply:
i tested your query and the time difference is not really measurable. but what happens indeed is that two more queries are fired now. one for retrieving all ips of all related objects and one for retrieving servertypes for the nested query.
at least django is smart enough to batch all servers together to retrieve a deferred column.
so while this adds unnecessary overhead i would like to not optimize it any further in this pr because i see the root problem in the design of the query logic.
fixing this properly imo requires rewriting the query classes in a way that they know what they are actually querying upfront, which will lead to more improvements where we could shrink down the total amount of queries made in general

Reply:
Cool! I agree as long as the deferred attribute queries get batched it doesn't really matter. :)